jwst/associations/lib/rules_level2b.py

```diff
@@ -502,55 +505,145 @@ def _init_hook(self, item):
+    def add_catalog_member(self):
+        """Add catalog member based on direct image members"""
+        directs = self.members_by_type('direct_image')
+        if not directs:
+            raise AssociationNotValidError(
+                '{} has no required direct image exposures'.format(
+                    self.__class__.__name__
+                )
+            )
 
-        # Get the Level3 product name of this association.
-        # Except for the grism component, it should be what
-        # the Level3 direct image name is.
+        sciences = self.members_by_type('science')
+        if not sciences:
+            raise AssociationNotValidError(
+                '{} has no required science exposure'.format(
+                    self.__class__.__name__
+                )
+            )
+        science = sciences[0]
+
+        # Get the exposure sequence for the science. Then, find
+        # the direct image greater than but closest to this value.
+        expspcin = int(getattr_from_list(science.item, ['expspcin'], _EMPTY)[1])
+        min_diff = -1         # Initialize to an invalid value.
+        closest = directs[0]  # If the search fails, just use the first.
+        for direct in directs:
+            direct_expspcin = int(getattr_from_list(
+                direct.item, ['expspcin'], _EMPTY
+            )[1])
+            diff = direct_expspcin - expspcin
+            if diff > min_diff:
+                min_diff = diff
+                closest = direct
+
+        # Note the selected direct image. Used in `Asn_Lv2WFSS._get_opt_element`
+        self.direct_image = closest
+
+        # Remove all other direct images from the association.
+        members = self.current_product['members']
+        direct_idxs = [
+            idx
+            for idx, member in enumerate(members)
+            if member['exptype'] == 'direct_image' and member != closest
+        ]
+        deque((
+            list.pop(members, idx)
+            for idx in sorted(direct_idxs, reverse=True)
+        ))
+
+        # Finally, add the catalog member
         lv3_direct_image_catalog = DMS_Level3_Base._dms_product_name(self) + '_cat.ecsv'
-
-        # Insert the needed catalog member
-        member = {
+        catalog_member = Member({
             'expname': lv3_direct_image_catalog,
             'exptype': 'sourcecat'
-        }
-        members = self.current_product['members']
-        members.append(member)
+        })
+        members.append(catalog_member)
+
+    def finalize(self):
+        """Finalize the association
+
+        For WFSS, this involves taking all the direct image exposures,
+        determine which one is first after last science exposure,
+        and creating the catalog name from that image.
+        """
+        try:
+            self.add_catalog_member()
+        except AssociationNotValidError:
+            logger.debug(
+                '%s has no required direct image exposures',
```

Comment:
Do we need this here? Doesn't line 569 catch this first?

Reply:
Though the catch is necessary since `finalize` returns None if no association is available. However, the logging is wrong; it should use the exception. Will modify.